modules/core/shared/src/main/scala/io/circe/JsonNumber.scala

```diff
@@ -259,7 +271,7 @@ object JsonNumber {
     else {
       val longValue = java.lang.Long.parseLong(value)
 
-      if (value.charAt(0) == '-' && longValue == 0L) JsonDecimal(value) else JsonLong(longValue)
+      if (value.charAt(0) == '-' && longValue == 0L) JsonDecimal(value) else fromLong(longValue)
```

Comment:
If you change the table to `JNumber` as suggested in another commented then you would also change here:

```suggestion
      if (value.charAt(0) == '-' && longValue == 0L) JsonDecimal(value) else fromLong(longValue).value
```